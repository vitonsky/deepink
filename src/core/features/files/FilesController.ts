import { z } from 'zod';
import { PGLiteDatabase } from '@core/storage/database/pglite/PGLiteDatabase';
import { qb } from '@utils/db/query-builder';
import { wrapDB } from '@utils/db/wrapDB';

import { IFilesStorage } from '.';

// TODO: add tests
// TODO: add runtime validation
// TODO: implement interface and use interface instead of class
/**
 * Files manager for local database
 */
export class FilesController {
	private db;
	private fileController;
	private readonly workspace;
	constructor(db: PGLiteDatabase, fileController: IFilesStorage, workspace: string) {
		this.db = db;
		this.fileController = fileController;
		this.workspace = workspace;
	}

	public async add(file: File) {
		const db = wrapDB(this.db.get());

		// Insert in DB
		const {
			rows: [{ id: fileId }],
		} = await db.query(
			qb.sql`INSERT INTO files (workspace_id,name,mimetype) VALUES (${this.workspace},${file.name},${file.type}) RETURNING id`,
			z.object({ id: z.string() }),
		);

		// TODO: delete entry in DB if can't upload file. Or try to upload first and then add file to DB
		// TODO: encrypt file
		// Write file
		const buffer = await file.arrayBuffer();
		await this.fileController.write(this.getFilePath(fileId), buffer);

		return fileId;
	}

	public async get(id: string) {
		const db = wrapDB(this.db.get());

		// Insert in DB
		const [fileEntry] = await db
			.query(
				qb.sql`SELECT * FROM files WHERE workspace_id=${this.workspace} AND id=${id}`,

				z.object({
					name: z.string(),
					mimetype: z.string(),
				}),
			)
			.then((result) => result.rows);

		if (!fileEntry) return null;

		const { name, mimetype } = fileEntry;

		const buffer = await this.fileController.get(this.getFilePath(id));
		if (!buffer) return null;

		// TODO: decrypt file
		return new File([buffer], name, { type: mimetype });
	}

	// TODO: remove attached files
	public async delete(filesId: string[]) {
		const db = wrapDB(this.db.get());

		if (filesId.length === 0) return;

		// Delete in database
		await db.query(
			qb.sql`DELETE FROM files WHERE workspace_id=${
				this.workspace
			} AND id IN (${qb.values(filesId)})`,
		);

		// Delete files
		await this.fileController.delete(
			filesId.map((filename) => this.getFilePath(filename)),
		);
	}

	public async query() {
		const db = wrapDB(this.db.get());

		return await db
			.query(
				qb.sql`SELECT id, name, mimetype FROM files WHERE workspace_id=${this.workspace}`,
				z.object({
					id: z.string(),
					name: z.string(),
					mimetype: z.string(),
				}),
			)
			.then(({ rows }) => rows);
	}

	private getFilePath(filename: string) {
		return [this.workspace, filename].join('/');
	}
}
