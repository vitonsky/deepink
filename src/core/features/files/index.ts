// TODO: add method `has`
/**
 * Interface for persistent files storage
 * File name is UUID, not a file system path
 *
 * For desktop app it may be FS or cloud service, for web it may be API or indexedDB
 */
export type IFilesStorage = {
	write: (uuid: string, buffer: ArrayBuffer) => Promise<void>;
	get: (uuid: string) => Promise<ArrayBuffer | null>;
	delete: (uuid: string[]) => Promise<void>;
	list: () => Promise<string[]>;
};

export type IFileController = {
	write: (buffer: ArrayBuffer) => Promise<void>;
	get: () => Promise<ArrayBuffer | null>;
	delete: () => Promise<void>;
};
