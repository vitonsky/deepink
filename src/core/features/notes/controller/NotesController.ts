/* eslint-disable camelcase */
import { Query } from 'nano-queries/core/Query';
import { z } from 'zod';
import { PGLiteDatabase } from '@core/storage/database/pglite/PGLiteDatabase';
import { qb } from '@utils/db/query-builder';
import { wrapDB } from '@utils/db/wrapDB';

import { INote, INoteContent, NoteId } from '..';
import {
	INotesController,
	NoteMeta,
	NotesControllerFetchOptions,
	NoteSortField,
} from '.';

const RowScheme = z
	.object({
		id: z.string(),
		title: z.string(),
		text: z.string(),
		created_at: z.number(),
		updated_at: z.number(),
		history_disabled: z.number(),
		visible: z.number(),
	})
	.transform(
		({
			id,
			title,
			text,
			created_at,
			updated_at,
			history_disabled,
			visible,
		}): INote => ({
			id,
			createdTimestamp: created_at,
			updatedTimestamp: updated_at,
			isSnapshotsDisabled: Boolean(history_disabled),
			isVisible: Boolean(visible),
			content: { title, text },
		}),
	);

function formatNoteMeta(meta: Partial<NoteMeta>) {
	return Object.fromEntries(
		Object.entries(meta).map((item): [string, string | number] => {
			const [key, value] = item as unknown as {
				[K in keyof NoteMeta]: [K, NoteMeta[K]];
			}[keyof NoteMeta];

			switch (key) {
				case 'isSnapshotsDisabled':
					return ['history_disabled', Number(value)];
				case 'isVisible':
					return ['visible', Number(value)];
			}
		}),
	);
}

function getFetchQuery(
	{
		select,
		workspace,
	}: {
		select: Query;
		workspace?: string;
	},
	{ limit, page, tags = [], meta, sort }: NotesControllerFetchOptions = {},
) {
	if (page !== undefined && page < 1)
		throw new TypeError('Page value must not be less than 1');

	const metaEntries = Object.entries(
		formatNoteMeta({
			isVisible: true,
			...meta,
		}),
	);

	const sortFieldMap: { [K in NoteSortField]: string } = {
		id: 'id',
		createdAt: 'created_at',
		updatedAt: 'updated_at',
	};

	return qb.line(
		qb.sql`SELECT ${select} FROM notes`,
		qb
			.where(workspace ? qb.sql`workspace_id=${workspace}` : undefined)
			.and(
				tags.length === 0
					? undefined
					: qb.sql`id IN (SELECT target FROM attached_tags WHERE source IN (${qb.values(
							tags,
					  )}))`,
			)
			.and(...metaEntries.map(([key, value]) => qb.sql`${qb.raw(key)} = ${value}`)),
		sort
			? qb.line(
					qb.sql`ORDER BY`,
					qb.set([
						sort
							? qb.line(
									sortFieldMap[sort.by],
									sort.order === 'desc' ? 'DESC' : 'ASC',
							  )
							: undefined,
					]),
			  )
			: undefined,
		limit ? qb.limit(limit) : undefined,
		page && limit ? qb.offset((page - 1) * limit) : undefined,
	);
}

/**
 * Synced notes registry
 */
export class NotesController implements INotesController {
	private db;
	private readonly workspace;
	constructor(db: PGLiteDatabase, workspace: string) {
		this.db = db;
		this.workspace = workspace;
	}

	public async getById(id: NoteId): Promise<INote | null> {
		const db = wrapDB(this.db.get());

		const {
			rows: [note],
		} = await db.query(
			qb.sql`SELECT * FROM notes WHERE workspace_id=${this.workspace} AND id=${id}`,
			RowScheme,
		);

		return note ?? null;
	}

	public async getLength(query: NotesControllerFetchOptions = {}): Promise<number> {
		const db = wrapDB(this.db.get());

		const {
			rows: [{ count }],
		} = await db.query(
			getFetchQuery(
				{ select: qb.sql`COUNT(*) as count`, workspace: this.workspace },
				query,
			),
			z.object({ count: z.number() }),
		);

		return count;
	}

	public async get(query: NotesControllerFetchOptions = {}): Promise<INote[]> {
		const db = wrapDB(this.db.get());

		const { rows } = await db.query(
			getFetchQuery({ select: qb.sql`*`, workspace: this.workspace }, query),
			RowScheme,
		);

		return rows;
	}

	public async add(note: INoteContent, meta?: Partial<NoteMeta>): Promise<NoteId> {
		const db = wrapDB(this.db.get());

		const creationTime = new Date().getTime();

		// Insert data
		const metaEntries = Object.entries(formatNoteMeta(meta ?? {}));

		const {
			rows: [id],
		} = await db.query(
			qb.sql`INSERT INTO notes (${qb.set([
				qb.sql`"workspace_id","title","text","created_at","updated_at"`,
				...(metaEntries ? metaEntries.map(([key]) => key) : []),
			])}) VALUES (${qb.values([
				this.workspace,
				note.title,
				note.text,
				creationTime,
				creationTime,
				...metaEntries.map(([_key, value]) => value),
			])}) RETURNING id`,
			z.object({ id: z.string() }).transform(({ id }) => id),
		);

		return id;
	}

	public async update(id: string, updatedNote: INoteContent) {
		const db = wrapDB(this.db.get());

		const updateTime = new Date().getTime();
		const result = await db.query(
			qb.line(
				'UPDATE notes SET',
				qb.values({
					title: updatedNote.title,
					text: updatedNote.text,
					// TODO: should we really update note by update meta?
					updated_at: updateTime,
				}),
				qb.sql`WHERE id=${id} AND workspace_id=${this.workspace}`,
			),
		);

		if (!result.affectedRows || result.affectedRows < 1) {
			throw new Error('Note did not updated');
		}
	}

	public async delete(ids: NoteId[]): Promise<void> {
		const db = wrapDB(this.db.get());

		const { affectedRows = 0 } = await db.query(
			qb.sql`DELETE FROM notes WHERE workspace_id=${
				this.workspace
			} AND id IN (${qb.values(ids)})`,
		);

		if (affectedRows !== ids.length) {
			console.warn(
				`Not match deleted entries length. Expected: ${ids.length}; Deleted: ${affectedRows}`,
			);
		}
	}

	public async updateMeta(ids: NoteId[], meta: Partial<NoteMeta>): Promise<void> {
		const db = wrapDB(this.db.get());

		const { affectedRows = 0 } = await db.query(
			qb.sql`UPDATE notes SET ${qb.values(formatNoteMeta(meta))}
				WHERE workspace_id=${this.workspace} AND id IN (${qb.values(ids)})`,
		);

		if (affectedRows !== ids.length) {
			console.warn(
				`Not match updated entries length. Expected: ${ids.length}; Updated: ${affectedRows}`,
			);
		}
	}
}
