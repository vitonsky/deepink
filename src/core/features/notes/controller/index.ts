import { INote, INoteContent, NoteId } from '..';

export type NoteMeta = {
	isSnapshotsDisabled: boolean;
	isVisible: boolean;
};

export type NoteSortField = 'id' | 'createdAt' | 'updatedAt';

export type NotesControllerFetchOptions = {
	/**
	 * Limit notes
	 *
	 * @default 100
	 */
	limit?: number;

	/**
	 * Page number start of 1
	 *
	 * Implementations must thrown error for values less than 1
	 *
	 * @default 1
	 */
	page?: number;

	/**
	 * Filter notes by tags
	 */
	tags?: string[];

	/**
	 * Filters by note meta info
	 */
	meta?: Partial<NoteMeta>;

	search?: {
		text: string;
	};

	/**
	 * Sorting options
	 */
	sort?: {
		by: NoteSortField;
		order?: 'desc' | 'asc';
	};
};

/**
 * Notes controller interface
 */
export interface INotesController {
	/**
	 * Get note by ID
	 */
	getById(id: NoteId): Promise<INote | null>;

	/**
	 * Get number of notes
	 */
	getLength(query?: NotesControllerFetchOptions): Promise<number>;

	/**
	 * Primary method to get notes filtered by parameters
	 */
	get(options?: NotesControllerFetchOptions): Promise<INote[]>;

	/**
	 * Create note and return unique id of new note
	 */
	add(note: INoteContent, meta?: Partial<NoteMeta>): Promise<NoteId>;

	/**
	 * Update note by unique id
	 */
	update(id: NoteId, updatedNote: INoteContent): Promise<void>;

	/**
	 * Update note meta information
	 */
	updateMeta(ids: NoteId[], meta: Partial<NoteMeta>): Promise<void>;

	/**
	 * Deletes all notes with specified ids
	 */
	delete(ids: NoteId[]): Promise<void>;

	/**
	 * Updates the status of specific notes
	 */
	updateStatus(ids: NoteId[], status: { deleted: boolean }): Promise<void>;
}
