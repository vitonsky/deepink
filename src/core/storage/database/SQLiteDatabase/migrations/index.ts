import { Database } from 'better-sqlite3';

export type MigrationsTarget = Database;

const migrations = [
	{
		version: 1,
		up: async (db: MigrationsTarget) => {
			db.exec(`CREATE TABLE "files" (
				"id"	TEXT NOT NULL UNIQUE,
				"name"	TEXT NOT NULL,
				"mimetype"	TEXT NOT NULL,
				PRIMARY KEY("id")
			)`);
		},
	},
	{
		version: 2,
		up: async (db: MigrationsTarget) => {
			db.exec(`CREATE TABLE "attachments" (
				"id"	TEXT NOT NULL UNIQUE,
				"file"	TEXT NOT NULL,
				"note"	TEXT NOT NULL,
				PRIMARY KEY("id")
			)`);
		},
	},
	{
		version: 3,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec('DROP TABLE "attachments"');
				db.exec(`CREATE TABLE "attachments" (
					"id"	TEXT NOT NULL UNIQUE,
					"file"	TEXT NOT NULL,
					"note"	TEXT NOT NULL,
					PRIMARY KEY("id")
					UNIQUE(file,note)
					)`);
			})();
		},
	},
	{
		version: 4,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(`CREATE TABLE "workspaces" (
					"id"	TEXT NOT NULL UNIQUE,
					"name"	TEXT NOT NULL,
					PRIMARY KEY("id")
					)`);
			})();
		},
	},
	{
		version: 5,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(`CREATE TABLE "note_versions" (
					"id" TEXT NOT NULL UNIQUE,
					"note_id" TEXT NOT NULL,
					"created_at" INTEGER NOT NULL DEFAULT 0,
					"title" TEXT NOT NULL,
					"text" TEXT NOT NULL,
					PRIMARY KEY("id"),
					FOREIGN KEY (note_id) REFERENCES notes(id) ON DELETE CASCADE
				)`);
			})();
		},
	},
	{
		version: 6,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(
					`ALTER TABLE "notes" ADD COLUMN "isSnapshotsDisabled" INTEGER NOT NULL DEFAULT 0;`,
				);
			})();
		},
	},
	{
		version: 7,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(
					`ALTER TABLE "notes" ADD COLUMN "isVisible" INTEGER NOT NULL DEFAULT 1;`,
				);
			})();
		},
	},
] as const;

export const latestSchemaVersion = migrations[migrations.length - 1].version;

export const migrateToLatestSchema = async (db: MigrationsTarget, verbose = false) => {
	let currentVersion = Number(db.pragma('main.user_version', { simple: true }));

	if (verbose) {
		console.log('DB version', currentVersion);
	}

	// TODO: ensure sorted versions
	for (const migration of migrations) {
		if (currentVersion >= migration.version) continue;

		const nextVersion = migration.version;

		await migration.up(db);
		db.pragma(`main.user_version = ${nextVersion};`);

		currentVersion = nextVersion;

		if (verbose) {
			console.log('DB version updated to', currentVersion);
		}
	}
};
