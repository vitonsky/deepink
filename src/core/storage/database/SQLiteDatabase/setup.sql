-- TODO: remove default value for `workspace_id`
CREATE TABLE "notes" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"title" TEXT NOT NULL,
	"text" TEXT NOT NULL,
	"creationTime" INTEGER NOT NULL DEFAULT 0,
	"lastUpdateTime" INTEGER NOT NULL DEFAULT 0,
	"isSnapshotsDisabled" INTEGER NOT NULL DEFAULT 0,
	PRIMARY KEY("id")
);
-- Note text snapshots
CREATE TABLE "note_versions" (
	"id" TEXT NOT NULL UNIQUE,
	"note_id" TEXT NOT NULL,
	"created_at" INTEGER NOT NULL DEFAULT 0,
	"title" TEXT NOT NULL,
	"text" TEXT NOT NULL,
	PRIMARY KEY("id"),
	FOREIGN KEY (note_id) REFERENCES notes(id) ON DELETE CASCADE
);
CREATE TABLE "files" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"name" TEXT NOT NULL,
	"mimetype" TEXT NOT NULL,
	PRIMARY KEY("id")
);
CREATE TABLE "attachments" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"file" TEXT NOT NULL,
	"note" TEXT NOT NULL,
	PRIMARY KEY("id") UNIQUE(file, note)
);
CREATE TABLE "tags" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"name" TEXT NOT NULL,
	"parent" TEXT,
	PRIMARY KEY("id") UNIQUE(name, parent)
);
CREATE TABLE "attachedTags" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"source" TEXT NOT NULL,
	"target" TEXT NOT NULL,
	PRIMARY KEY("id") UNIQUE(source, target)
);
CREATE TABLE "workspaces" (
	"id" TEXT NOT NULL UNIQUE,
	"name" TEXT NOT NULL,
	PRIMARY KEY("id")
);