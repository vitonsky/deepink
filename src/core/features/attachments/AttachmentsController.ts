import { v4 as uuid4 } from 'uuid';
import { z } from 'zod';
import { qb } from '@utils/db/query-builder';
import { wrapDB } from '@utils/db/wrapDB';

import { SQLiteDatabase } from '../../storage/database/SQLiteDatabase/SQLiteDatabase';

/**
 * attachments manager, to track attachments usage and to keep consistency
 */
export class AttachmentsController {
	private db;
	private readonly workspace;
	constructor(db: SQLiteDatabase, workspace: string) {
		this.db = db;
		this.workspace = workspace;
	}

	public async set(targetId: string, attachments: string[]) {
		const db = this.db.get();

		if (attachments.length === 0) {
			db.prepare('DELETE FROM attachments WHERE workspace_id=? AND note=?').run(
				this.workspace,
				targetId,
			);
		} else {
			const valuesPlaceholders = attachments.map(() => `(?, ?, ?, ?)`).join(', ');

			const placeholdersData: string[] = [];
			attachments.forEach((attachmentId) => {
				placeholdersData.push(uuid4());
				placeholdersData.push(this.workspace);
				placeholdersData.push(targetId);
				placeholdersData.push(attachmentId);
			});

			db.transaction(() => {
				db.prepare('DELETE FROM attachments WHERE workspace_id=? AND note=?').run(
					this.workspace,
					targetId,
				);
				db.prepare(
					`INSERT INTO attachments ("id", "workspace_id", "note", "file") VALUES ${valuesPlaceholders};`,
				).run(placeholdersData);
			})();
		}
	}

	public async get(targetId: string): Promise<string[]> {
		const db = this.db.get();

		return db
			.prepare(
				'SELECT file FROM attachments WHERE workspace_id=? AND note=? ORDER BY rowid',
			)
			.all(this.workspace, targetId)
			.map((row) => (row as { file: string }).file);
	}

	public async delete(resources: string[]) {
		const db = this.db.get();

		const placeholders = Array(resources.length).fill('?').join(',');
		db.prepare(
			`DELETE FROM attachments WHERE workspace_id=? AND file IN (${placeholders})`,
		).run(this.workspace, ...resources);
	}

	public async query() {
		const db = wrapDB(this.db.get());

		return z
			.object({
				id: z.string(),
				file: z.string(),
				note: z.string(),
			})
			.array()
			.parse(
				db.all(
					qb.sql`SELECT id, file, note FROM attachments WHERE workspace_id=${this.workspace}`,
				),
			);
	}
}
