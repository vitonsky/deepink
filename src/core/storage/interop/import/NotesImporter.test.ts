import { AttachmentsController } from '@core/features/attachments/AttachmentsController';
import { createFileManagerMock } from '@core/features/files/__tests__/mocks/createFileManagerMock';
import { FilesController } from '@core/features/files/FilesController';
import { NotesController } from '@core/features/notes/controller/NotesController';
import { TagsController } from '@core/features/tags/controller/TagsController';
import { createFileControllerMock } from '@utils/mocks/fileControllerMock';

import { openDatabase } from '../../database/SQLiteDatabase/SQLiteDatabase';

import { NotesImporter } from '.';

const FAKE_WORKSPACE_NAME = 'fake-workspace-id';

const createTextBuffer = (text: string): ArrayBuffer => new TextEncoder().encode(text);

const dbFile = createFileControllerMock();
const dbPromise = openDatabase(dbFile);

afterAll(async () => {
	const db = await dbPromise;
	await db.close();
});

describe('Text buffers may be imported', () => {
	const fileManager = createFileManagerMock();

	test('Import notes', async () => {
		const db = await dbPromise;
		const notesRegistry = new NotesController(db, FAKE_WORKSPACE_NAME);
		const tagsRegistry = new TagsController(db, FAKE_WORKSPACE_NAME);

		const attachmentsRegistry = new AttachmentsController(db, FAKE_WORKSPACE_NAME);
		const filesRegistry = new FilesController(
			db,
			fileManager,
			attachmentsRegistry,
			FAKE_WORKSPACE_NAME,
		);

		const importer = new NotesImporter({
			filesRegistry,
			attachmentsRegistry,
			notesRegistry,
			tagsRegistry,
		});

		await importer.import({
			// Resources is a special directory with attachments
			'/_resources/secret.txt': createTextBuffer('SECRET'),
			// This files must not be imported,
			// since it is unused resources
			'/_resources/unused1': createTextBuffer('UNUSED FILE'),
			'/_resources/unused2.txt': createTextBuffer('UNUSED FILE'),
			'/_resources/unused3.md': createTextBuffer('UNUSED FILE'),

			// Notes list
			'/note-1.md': createTextBuffer('Hello world!'),
			'/note-2.md': createTextBuffer(
				'---\ntitle: Title from meta\n---\nHello world!',
			),
			'/note-3.md': createTextBuffer('Mention for [note #1](./note-1.md)'),
			'/note-4.md': createTextBuffer(
				'Text with [attachment](./_resources/secret.txt)',
			),
			'/note-5.md': createTextBuffer(
				'Another text with [attachment](./_resources/secret.txt)',
			),
			// Notes in subdirectories
			'/note-6.md': createTextBuffer(
				'Mention for non exist notes [one](./non-exist-note-1.md), [two](<./non exist note 2.md>), [three](./non/exist/note.md), [four](../../non-exist-note.md). And non exists attachments [one](./non-exist-file-1.jpeg), [two](<./non exist file 2.jpeg>), [three](./non/exist/file.jpeg), [four](../non-exist-file.jpeg)',
			),
			'/note-7.md': createTextBuffer(
				'Mention for [note #1](/note-1.md) via absolute path',
			),
			// Notes in subdirectories
			'/section-1/note #1.md': createTextBuffer('Hello world!'),
			'/section-1/note #2.md': createTextBuffer(
				'Mention for [note #1](<./note #1.md>)',
			),

			// Subdirectories with resources
			'/posts/about/image.png': createTextBuffer('IMAGE CONTENT HERE'),
			'/posts/about/unused.png': createTextBuffer('UNUSED IMAGE CONTENT HERE'),
			'/posts/about/index.md': createTextBuffer(
				'---\ntitle: Blog post\n---\nPost with image\n\n![attached image](./image.png)\nThis post and all its resources is placed in subdirectory.',
			),
		});
	});

	test('Imported notes is in list', async () => {
		const db = await dbPromise;
		const notesRegistry = new NotesController(db, FAKE_WORKSPACE_NAME);

		await expect(notesRegistry.get()).resolves.toEqual([
			expect.objectContaining({
				// TODO: fix unnecessary new line
				content: { title: 'note-1', text: 'Hello world!\n' },
			}),
			expect.objectContaining({
				content: { title: 'Title from meta', text: 'Hello world!\n' },
			}),
			expect.objectContaining({
				content: {
					title: 'note-3',
					text: expect.stringMatching(
						/^Mention for \[note #1\]\(note:\/\/[\d\w-]+\)\n$/,
					),
				},
			}),
			expect.objectContaining({
				content: {
					title: 'note-4',
					text: expect.stringMatching(
						/^Text with \[attachment\]\(res:\/\/[\d\w-]+\)\n$/,
					),
				},
			}),
			expect.objectContaining({
				content: {
					title: 'note-5',
					text: expect.stringMatching(
						/^Another text with \[attachment\]\(res:\/\/[\d\w-]+\)\n$/,
					),
				},
			}),
			expect.objectContaining({
				content: {
					title: 'note-6',
					text: 'Mention for non exist notes [one](./non-exist-note-1.md), [two](<./non exist note 2.md>), [three](./non/exist/note.md), [four](../../non-exist-note.md). And non exists attachments [one](./non-exist-file-1.jpeg), [two](<./non exist file 2.jpeg>), [three](./non/exist/file.jpeg), [four](../non-exist-file.jpeg)\n',
				},
			}),

			expect.objectContaining({
				content: {
					title: 'note-7',
					text: expect.stringMatching(
						/^Mention for \[note #1\]\(note:\/\/[\d\w-]+\) via absolute path\n$/,
					),
				},
			}),
			expect.objectContaining({
				// TODO: fix unnecessary new line
				content: { title: 'note #1', text: 'Hello world!\n' },
			}),
			expect.objectContaining({
				content: {
					title: 'note #2',
					text: expect.stringMatching(
						/^Mention for \[note #1\]\(note:\/\/[\d\w-]+\)\n$/,
					),
				},
			}),
			expect.objectContaining({
				content: {
					title: 'Blog post',
					text: expect.stringMatching(
						/^Post with image\n\n!\[attached image\]\(res:\/\/[\d\w-]+\)\nThis post and all its resources is placed in subdirectory.\n$/,
					),
				},
			}),
		]);
	});

	test('Only attached files must be saved on FS', async () => {
		// If file used in many places, it must have many references,
		// but physically we have only one file
		await expect(fileManager.list()).resolves.toHaveLength(2);
	});

	test('Attached files is in files list', async () => {
		const db = await dbPromise;
		const notesRegistry = new NotesController(db, FAKE_WORKSPACE_NAME);

		const note = await notesRegistry.get().then((notes) => {
			const note = notes.find((note) => note.content.title === 'note-4');
			if (!note) throw new Error('Not found note by title');

			return note;
		});

		const attachmentsRegistry = new AttachmentsController(db, FAKE_WORKSPACE_NAME);
		const filesRegistry = new FilesController(
			db,
			fileManager,
			attachmentsRegistry,
			FAKE_WORKSPACE_NAME,
		);

		const attachmentIds = await attachmentsRegistry.get(note.id);
		expect(attachmentIds).toHaveLength(1);

		await expect(
			filesRegistry
				.get(attachmentIds[0])
				.then((file) => (file ? file.text() : null)),
		).resolves.toEqual('SECRET');
	});

	test('Tags is created and reproduces structure in FS', async () => {
		const db = await dbPromise;
		const tagsRegistry = new TagsController(db, FAKE_WORKSPACE_NAME);

		await expect(tagsRegistry.getTags()).resolves.toEqual([
			expect.objectContaining({
				name: 'section-1',
				parent: null,
				resolvedName: 'section-1',
			}),
			expect.objectContaining({
				name: 'posts',
				parent: null,
				resolvedName: 'posts',
			}),
			expect.objectContaining({
				name: 'about',
				parent: expect.stringMatching(/^[\d\w-]+$/),
				resolvedName: 'posts/about',
			}),
		]);
	});
});
