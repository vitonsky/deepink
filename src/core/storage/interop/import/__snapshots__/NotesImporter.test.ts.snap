// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Base notes import cases > Import notes > onProcessed hook calls 1`] = `
[
  [
    {
      "processed": 1,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 2,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 3,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 4,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 5,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 6,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 7,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 8,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 9,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 10,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 11,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 12,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 13,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 14,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 15,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 16,
      "stage": "parsing",
      "total": 16,
    },
  ],
  [
    {
      "processed": 1,
      "stage": "uploading",
      "total": 2,
    },
  ],
  [
    {
      "processed": 2,
      "stage": "uploading",
      "total": 2,
    },
  ],
  [
    {
      "processed": 1,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 2,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 3,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 4,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 5,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 6,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 7,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 8,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 9,
      "stage": "updating",
      "total": 10,
    },
  ],
  [
    {
      "processed": 10,
      "stage": "updating",
      "total": 10,
    },
  ],
]
`;
